pkg/cloudprovider/providers/vsphere/vsphere.go

```diff
@@ -701,835 +446,193 @@ func (vs *VSphere) ScrubDNS(nameservers, searches []string) (nsOut, srchOut []st
+		return nil, err
 	}
-	return nil
-}
-
-// Attaches given virtual disk volume to the compute running kubelet.
-func (vs *VSphere) AttachDisk(vmDiskPath string, storagePolicyID string, nodeName k8stypes.NodeName) (diskID string, diskUUID string, err error) {
-	attachDiskInternal := func(vmDiskPath string, storagePolicyID string, nodeName k8stypes.NodeName) (diskID string, diskUUID string, err error) {
-		var newSCSIController types.BaseVirtualDevice
-
-		// Create context
-		ctx, cancel := context.WithCancel(context.Background())
-		defer cancel()
-
-		// Ensure client is logged in and session is valid
-		err = vSphereLogin(ctx, vs)
-		if err != nil {
-			glog.Errorf("Failed to login into vCenter - %v", err)
-			return "", "", err
-		}
-
-		// Find virtual machine to attach disk to
-		var vSphereInstance string
-		if nodeName == "" {
-			vSphereInstance = vs.localInstanceID
-			nodeName = vmNameToNodeName(vSphereInstance)
+	if !nodeExist {
+		glog.Errorf("DisksAreAttached failed to determine whether disks %v are still attached: node %q does not exist",
+			volPaths,
+			vSphereInstance)
+		return nil, fmt.Errorf("DisksAreAttached failed to determine whether disks %v are still attached: node %q does not exist",
+			volPaths,
+			vSphereInstance)
+	}
+	// Create context
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	vm, err := vs.getVMByName(ctx, nodeName)
+	if err != nil {
+		glog.Errorf("Failed to get VM object for node: %q. err: +%v", nodeNameToVMName(nodeName), err)
+		return nil, err
+	}
+	for _, volPath := range volPaths {
+		result, err := vm.IsDiskAttached(ctx, volPath)
+		if err == nil {
+			if result {
+				attached[volPath] = true
+			} else {
+				attached[volPath] = false
+			}
 		} else {
-			vSphereInstance = nodeNameToVMName(nodeName)
+			return nil, err
 		}
+	}
+	return attached, nil
+}
 
-		// Get VM device list
-		vm, vmDevices, dc, err := getVirtualMachineDevices(ctx, vs.cfg, vs.client, vSphereInstance)
+// CreateVolume creates a volume of given size (in KiB).
+func (vs *VSphere) CreateVolume(volumeOptions *vclib.VolumeOptions) (volumePath string, err error) {
+	var datastore string
+	// Default datastore is the datastore in the vSphere config file that is used to initialize vSphere cloud provider.
+	if volumeOptions.Datastore == "" {
+		datastore = vs.cfg.Global.Datastore
+	} else {
+		datastore = volumeOptions.Datastore
+	}
+	// Ensure client is logged in and session is valid
+	err = vs.conn.Connect()
+	if err != nil {
+		return "", err
+	}
+	// Create context
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	dc, err := vclib.GetDatacenter(ctx, vs.conn, vs.cfg.Global.Datacenter)
+	if err != nil {
+		return "", err
+	}
+	var vmOptions *vclib.VMOptions
+	if volumeOptions.VSANStorageProfileData != "" || volumeOptions.StoragePolicyName != "" {
+		// Acquire a read lock to ensure multiple PVC requests can be processed simultaneously.
```

Comment:
I think taking lock for dummyVM should be in the diskManager.vmdm.go
Why caller needs to take this lock?

Because of this any one using vclib need to synchronize using some outside locking scheme